admin-bulk-user-membership/tool/src/main/java/org/sakaiproject/bulk/membership/controller/MainController.java

```diff
@@ -0,0 +1,247 @@
+    private String[] roles = null;
+    private ArrayList<Site> sites = new ArrayList<Site>();
+
+    private int currentStep = 1;
+    private String action = "";
+
+    @ModelAttribute("locale")
+	public Locale localeResolver(HttpServletRequest request, HttpServletResponse response) {
+        Locale loc = bulkUserMembershipToolService.getLocaleForCurrentUser();
+        LocaleResolver localeResolver = RequestContextUtils.getLocaleResolver(request);
+        localeResolver.setLocale(request, response, loc);
+        return loc;
+    }
+
+    @GetMapping(value = {"/", "/index"})
+    public String index(Model model) {
+        log.info("Starting the Bulk User Membership tool");
+        startAgain(model);
+        return usePage(model);
+    }
+
+    @GetMapping(value = {"/step/{step}", "/auto_membership_step_1", "/auto_membership_step_1_5", "/auto_membership_step_2", "/auto_membership_step_3"})
+    public String usePage(Model model) {
+        String nextPage = "";
+        model.addAttribute("users", StringUtils.join(users, "\n"));
+
+        String[] sitesIds = new String[sites.size()];
+        for (int index = 0; index < sites.size(); index++) {
+            sitesIds[index] = sites.get(index).getId();
+        }
+
+        model.addAttribute("sitesIds", StringUtils.join(sitesIds, "\n"));
+        model.addAttribute("action", action);
+        switch (currentStep) {
+            case 0:
+                model.addAttribute("roles", roles);
+                model.addAttribute("sites", sites);
+                nextPage = AUTO_MEMBERSHIP_STEP_1_5;
+                break;
+            case 1:
+                nextPage = AUTO_MEMBERSHIP_STEP_1;
+                break;
+            case 2:
+                nextPage = AUTO_MEMBERSHIP_STEP_2;
+                break;
+            case 3:
+                nextPage = AUTO_MEMBERSHIP_STEP_3;
+                break;
+        }
+        return nextPage;
+    }
+
+    @PostMapping(value = {"/step/1", "/auto_membership_step_1"})
+    public String checkFirstStep(@RequestParam(value="users", required = false) String usersCriteria, 
+                                 @RequestParam(value="sites", required = false) String sitesIds,
+                                 @RequestParam(value="action", required = false) String action,
+                                 RedirectAttributes redirectAttributes, Model model) {
+
+        if (usersCriteria.isEmpty() || sitesIds.isEmpty() || (action==null || action.isEmpty())) {
+            redirectAttributes.addFlashAttribute("noUsers", usersCriteria.isEmpty());
+            redirectAttributes.addFlashAttribute("noSites", sitesIds.isEmpty());
+            redirectAttributes.addFlashAttribute("noAction", (action==null || action.isEmpty()));
+            redirectAttributes.addFlashAttribute("users", usersCriteria);
+            redirectAttributes.addFlashAttribute("sites", sitesIds);
+            return "redirect:/" + AUTO_MEMBERSHIP_STEP_1;
```

Comment:
Create a constant for "redirect:/

private final String REDIRECT = ""redirect:/";

............

return REDIRECT + AUTO_MEMBERSHIP_STEP_1;